src/decoders/belief_propogation.jl

```diff
@@ -0,0 +1,139 @@
+using SparseArrays
+
+include("abstract_decoder.jl")
+
+struct BeliefPropagationSetup
+  "Log probabilities"
+  log_probabs::Vector{Float64}
+  
+  "Channel probabilities"
+  channel_probs::Vector{Float64}
+  
+  "Bit to check message matrix"
+  bit_2_check::Matrix{Float64}
+
+  "Check to bit message matrix"
+  check_2_bit::Matrix{Float64}
+
+  "predicted error for each bit"
+  err::Vector{Float64}
+end 
+
+
+struct BeliefPropagationDecoder{T} <: AbstractDecoder
+  "Parity check matrix"
+  H::T
+
+  "Physical error rate"
+  per::Float64
+
+  "Number of max iterations of Belief propagation decoder"
+  max_iters::Int
+end
+
+struct BeliefPropagationDecoderSparse{T} <: AbstractDecoder
+  "Parity check matrix"
+  H::T
+
+  "Physical error rate"
+  per::Float64
+
+  "Number of max iterations of Belief propagation decoder"
+  max_iters::Int
+  
+  "Num of parity checks i.e. number of rows of parity check matrix"
+  s::Int
+  
+  "Num of bits in the code i.e number of columns of parity check matrix"
+  n::Int
+  
+  sparse_H::SparseArrays.SparseMatrixCSC{Bool,Int}
+  
+  sparse_HT::SparseArrays.SparseMatrixCSC{Bool,Int}
+end
+
+"""
+Function to reset the scratch space for Belief propagation decoding
+
+# Arguments
+* `bp_setup`: The Belief Propagation Scratch space
+* `bp_decoder`: The belief propagation decoder configuration
+
+# Examples
+julia> decode!(bpd, syndrome, errors)
```

Comment:
Could you make the examples into `jldoctests`? That way they will be executed as part of the tests and we will know they have not been broken.